### PR TITLE
Rename StoreVectorMxNAndZip to StoreVectorAndZip

### LINKED ISSUE #102340: Arm64: Rename AdvSimd/Sve Load/Store vector API names
### AdvSimd
1. Rename `LoadVector128xN` to `LoadNxVector128 `
   -  Consistent naming. See https://github.com/dotnet/runtime/issues/97831#issuecomment-1934714477
2. Rename `StoreVectorNxMAndZip` to `StoreVectorAndZip`
   - The type on which the API is operation on can be known from parameters. As such we do not need special naming for them.
3. Rename `StoreVectorNx2` to `Store`
   - Same reason as point 2) above.

### Sve
1. Rename `LoadNxVector` to `LoadNxAndUnzip`
   - The operation does the unzip, similar to how it is done in AdvSimd. Compare left side operation vs. "4th quadrant" of right side.
image

2. Rename `Store` to `StoreAndZip`
   - Same reason as point 1) above

Discussion: https://github.com/dotnet/runtime/pull/102180#discussion_r1602562814

### PR DESCRIPTION
**Pull request by SwapnilGaikwad**
Partially (item 2) fixes: #102340

### CODE REVIEW

**Comment by SwapnilGaikwad** [comment]
Hi @kunalspathak, the coreclr part of this PR is tested but the mono bit is not. There are two commits separating the coreclr and mono changes that would make it easier to undo/update mono changes if needed.

File: src/coreclr/jit/hwintrinsiclistarm64.h [hunk 1]
```diff
@@ -469,6 +469,7 @@ HARDWARE_INTRINSIC(AdvSimd,       StoreSelectedScalar,
 HARDWARE_INTRINSIC(AdvSimd,       StoreSelectedScalarVector64x2,                                     8,      3,      true,  {INS_st2,            INS_st2,            INS_st2,            INS_st2,            INS_st2,            INS_st2,            INS_invalid,        INS_invalid,        INS_st2,            INS_invalid},     HW_Category_MemoryStore,           HW_Flag_BaseTypeFromFirstArg|HW_Flag_HasImmediateOperand|HW_Flag_SIMDScalar|HW_Flag_SpecialCodeGen|HW_Flag_SpecialImport|HW_Flag_NeedsConsecutiveRegisters)
```
Comment (kunalspathak) [inline src/coreclr/jit/hwintrinsiclistarm64.h:469]: wondering if we need `StoreSelectedScalarVector64x*` and `StoreSelectedScalarVector128x*`? Can't we know that from the tuple size of arguments and use that find out instruction?

File: src/coreclr/jit/hwintrinsiclistarm64.h [hunk 2]
```diff
@@ -469,6 +469,7 @@ HARDWARE_INTRINSIC(AdvSimd,       StoreSelectedScalar,
...
 HARDWARE_INTRINSIC(AdvSimd,       StoreSelectedScalarVector64x4,                                     8,      3,      true,  {INS_st4,            INS_st4,            INS_st4,            INS_st4,            INS_st4,            INS_st4,            INS_invalid,        INS_invalid,        INS_st4,            INS_invalid},     HW_Category_MemoryStore,           HW_Flag_BaseTypeFromFirstArg|HW_Flag_HasImmediateOperand|HW_Flag_SIMDScalar|HW_Flag_SpecialCodeGen|HW_Flag_SpecialImport|HW_Flag_NeedsConsecutiveRegisters)
+HARDWARE_INTRINSIC(AdvSimd,       StoreVectorAndZip,                                                 8,      2,     false,  {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid},     HW_Category_MemoryStore,           HW_Flag_BaseTypeFromFirstArg|HW_Flag_SpecialImport|HW_Flag_SpecialCodeGen)
```
Comment (kunalspathak) [inline src/coreclr/jit/hwintrinsiclistarm64.h:472]: As mentioned above, with `SpecialCodeGen` flag for `StoreVectorAndZip`, we should be able to eliminate `StoreSelectedScalarVector64x*` and `StoreSelectedScalarVector128x*`.

File: src/coreclr/jit/hwintrinsicarm64.cpp [hunk 3]
```diff
@@ -2042,12 +2042,69 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
...
+                impStoreToTemp(tmp, op2, CHECK_SPILL_NONE);
+                op2 = gtNewLclvNode(tmp, argType);
+            }
+            op2 = gtConvertTableOpToFieldList(op2, fieldCount);
+
+            switch (fieldCount)
+            {
+                case 2:
+                    intrinsic =
+                        simdSize == 8 ? NI_AdvSimd_StoreVector64x2AndZip : NI_AdvSimd_Arm64_StoreVector128x2AndZip;
```
Comment (kunalspathak) [inline src/coreclr/jit/hwintrinsicarm64.cpp:2086]: same here

File: src/coreclr/jit/hwintrinsiclistarm64.h (hunk 2 above)
Comment (kunalspathak) [reply src/coreclr/jit/hwintrinsiclistarm64.h:472]: Something similar to `NI_AdvSimd_VectorTableLookup`

**Review by kunalspathak — CHANGES_REQUESTED** [review]
Added comment to handle it similar to `NI_AdvSimd_VectorTableLookup`

**Comment by kunalspathak** [comment]
@fanyang-mono - which CI pipelines we should run to make sure that the code in the PR is getting tested for mono?

File: src/coreclr/jit/hwintrinsiclistarm64.h (hunk 2 above)
Comment (SwapnilGaikwad) [reply src/coreclr/jit/hwintrinsiclistarm64.h]: Done

File: src/coreclr/jit/hwintrinsicarm64.cpp (hunk 3 above)
Comment (SwapnilGaikwad) [reply src/coreclr/jit/hwintrinsicarm64.cpp]: Done

**Comment by SwapnilGaikwad** [comment]
> Added comment to handle it similar to `NI_AdvSimd_VectorTableLookup`

The `StoreVectorAndZip` is now cleaned-up and matches with `VectorTableLookup`. On the similar note but unrelated to `StoreAndZip` - we also got `StoreSelectedScalarNxM` that can be cleaned-up similarly. 
Would you like me to throw those changes in this PR or a separate one?

**Review by kunalspathak — COMMENTED** [review]
looks much simpler. Added few last comments.

**Comment by kunalspathak** [comment]
> Would you like me to throw those changes in this PR or a separate one?

I am ok with either way. Really depends on if you have cycles to do that change here, or later.

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp [hunk 4]
```diff
@@ -1307,6 +1301,49 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
...
+                GenTreeFieldList* fieldList  = intrin.op2->AsFieldList();
+                GenTree*          firstField = fieldList->Uses().GetHead()->GetNode();
+                op2Reg                       = firstField->GetRegNum();
+
+                regNumber argReg = op2Reg;
+                for (GenTreeFieldList::Use& use : fieldList->Uses())
+                {
+                    regCount++;
+
+                    GenTree* argNode = use.GetNode();
```
[the remark this answers is not in the document]
Comment (kunalspathak) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:1320]: fixed this.

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp [hunk 5]
```diff
@@ -1307,6 +1301,49 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
...
+
+                    case 4:
+                        ins = INS_st4;
+                        break;
+
+                    default:
+                        unreached();
+                }
+
+                GetEmitter()->emitIns_R_R(ins, emitSize, op2Reg, op1Reg, opt);
```
[the remark this answers is not in the document]
Comment (SwapnilGaikwad) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:1344]: Sure, it's asserted in `genGetSimdInsOpt()` one line [322](https://github.com/dotnet/runtime/pull/103638/files/92a7b53128d4d4a2ae08ef966726309bf89611c8..c7ce1bdb75e8b3757de1c843000f8efcca5ab65a#diff-45c9473f1b1d031652a13186a211c1aef1406a74f576d5f04e3a65766b0f5050L322) above. 
We aren't asserting explicitly in other cases, is there a chance getting other values by any chance? 
For `HW_Category_MemoryStore`, the `emitSize` seemed to be pulled from the table in `hwintrinsiclistarm64.h.`

[the remark this answers is not in the document]
Comment (kunalspathak) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:1344]: yes, that's fine.

**Review by kunalspathak — APPROVED** [review]
LGTM coreclr side

**Comment by fanyang-mono** [comment]
Test `AdvSimd_r.dll` passed locally

```
Supported ISAs:
  AdvSimd:   True
  Aes:       True
  ArmBase:   True
  Crc32:     True
  Dp:        False
  Rdm:       False
  Sha1:      True
  Sha256:    True
  Sve:       False

...

Expected: 100
Actual: 100
END EXECUTION - PASSED
```

**Review by fanyang-mono — APPROVED** [review]
Mono change looks good to me.

**Comment by kunalspathak** [comment]
/ba-g failures are known or mono amd64, unrelated

**Pull request merged**